akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/DynamicShardAllocationSpec.scala

```diff
@@ -0,0 +1,189 @@
+  val forth = role("forth")
+}
+
+class DynamicShardAllocationSpecMultiJvmNode1 extends DynamicShardAllocationSpec
+class DynamicShardAllocationSpecMultiJvmNode2 extends DynamicShardAllocationSpec
+class DynamicShardAllocationSpecMultiJvmNode3 extends DynamicShardAllocationSpec
+class DynamicShardAllocationSpecMultiJvmNode4 extends DynamicShardAllocationSpec
+
+object DynamicShardAllocationSpec {
+
+  object GiveMeYourHome {
+    case class Get(id: String)
+    case class Home(address: Address)
+
+    val extractEntityId: ShardRegion.ExtractEntityId = {
+      case g @ Get(id) => (id, g)
+    }
+
+    // shard == id to make testing easier
+    val extractShardId: ShardRegion.ExtractShardId = {
+      case Get(id) => id
+    }
+  }
+
+  class GiveMeYourHome extends Actor with ActorLogging {
+
+    val selfAddress = Cluster(context.system).selfAddress
+
+    log.info("Started on {}", selfAddress)
+
+    override def receive: Receive = {
+      case Get(_) =>
+        sender() ! Home(selfAddress)
+    }
+  }
+}
+
+abstract class DynamicShardAllocationSpec
+    extends MultiNodeSpec(DynamicShardAllocationSpecConfig)
+    with MultiNodeClusterSpec
+    with ImplicitSender
+    with ScalaFutures {
+
+  import DynamicShardAllocationSpecConfig._
+  import DynamicShardAllocationSpec._
+  import DynamicShardAllocationSpec.GiveMeYourHome._
+
+  override implicit val patienceConfig: PatienceConfig = PatienceConfig(5.second)
+
+  val typeName = "home"
+  val initiallyOnForth = "on-forth"
+
+  "Dynamic shard allocation" must {
+    "form cluster" in {
+      awaitClusterUp(first, second, third)
+      enterBarrier("cluster-started")
+    }
+
+    lazy val shardRegion = {
+      ClusterSharding(system).start(
+        typeName = typeName,
+        entityProps = Props[GiveMeYourHome],
+        settings = ClusterShardingSettings(system),
+        extractEntityId = extractEntityId,
+        extractShardId = extractShardId,
+        () => new DynamicShardAllocationStrategy(system, typeName),
+        PoisonPill)
+    }
+
+    "start cluster sharding" in {
+      shardRegion
+      enterBarrier("shard-region-started")
+    }
+
+    "default to allocating a shard to the local shard region" in {
+      runOn(first, second, third) {
+        shardRegion ! Get(myself.name)
+        val actorLocation = expectMsgType[Home](20.seconds).address
+        actorLocation shouldEqual Cluster(system).selfAddress
+      }
+      enterBarrier("local-message-sent")
+    }
+
+    "move shard via distributed data" in {
+      val shardToSpecifyLocation = "cats"
+      runOn(first) {
+        DynamicShardAllocation(system)
+          .clientFor(typeName)
+          .updateShardLocation(shardToSpecifyLocation, Cluster(system).selfAddress)
+          .futureValue
+      }
+      enterBarrier("shard-location-updated")
+
+      runOn(second, third) {
+        awaitAssert({
+          shardRegion ! Get(shardToSpecifyLocation)
+          expectMsg(Home(address(first)))
```

Comment:
There is a risk that an extra reply ends up in the `testActor` and fails next test step when using it like this inside awaitAssert. Safer to create a separate probe.

Reply:
:+1: